sql/core/src/test/scala/org/apache/spark/sql/PlanStabilitySuite.scala

```diff
@@ -0,0 +1,304 @@
+import org.apache.spark.sql.execution._
+import org.apache.spark.sql.execution.adaptive.DisableAdaptiveExecutionSuite
+import org.apache.spark.sql.execution.exchange.{Exchange, ReusedExchangeExec}
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * Check that TPCDS SparkPlans don't change.
+ * If there is a regression, the error message looks like this:
+ *   Plans did not match:
+ *   last approved plan: /path/to/tpcds-plan-stability/approved-plans-xxx/q1/0.simplified.txt
+ *   last explain: /path/to/tpcds-plan-stability/approved-plans-xxx/q1/0.explain.txt
+ *   actual plan: /path/to/tmp/q1.actual.simplified.txt
+ *   actual explain: /path/to/tmp/q1.actual.explain.txt
+ *   [side by side plan diff]
+ * The explain files are saved to help debug later, they are not checked. Only the simplified
+ * plans are checked (by string comparison).
+ *
+ * Approving new plans:
+ * IF the plan change is intended then re-running the test
+ * with environ var SPARK_GENERATE_GOLDEN_FILES=1 will make the new plan canon.
+ * This should be done only for the queries that need it, to avoid unnecessary diffs in the
+ * other approved plans.
+ * This can be done by running sbt test-only *PlanStabilitySuite* -- -z "(q31)"
+ * The new plan files should be part of the PR and reviewed.
+ *
+ * Multiple approved plans:
+ * It's possible that a query has multiple correct plans. This should be decided as part of the
+ * review. In this case, change the call to approvePlans and set variant=true.
+ */
+trait PlanStabilitySuite extends TPCDSBase with DisableAdaptiveExecutionSuite {
+
+  private val regenerateGoldenFiles: Boolean = System.getenv("SPARK_GENERATE_GOLDEN_FILES") == "1"
+
+  protected val baseResourcePath = {
+    // If regenerateGoldenFiles is true, we must be running this in SBT and we use hard-coded
+    // relative path. Otherwise, we use classloader's getResource to find the location.
+    if (regenerateGoldenFiles) {
+      java.nio.file.Paths.get("src", "test", "resources", "tpcds-plan-stability").toFile
+    } else {
+      val res = getClass.getClassLoader.getResource("tpcds-plan-stability")
+      new File(res.getFile)
+    }
+  }
+
+  def goldenFilePath: String
+
+  private def getExistingVariants(name: String): Array[Int] = {
+    val dir = new File(goldenFilePath, s"$name")
+    // file paths are the form ../q3/2.simplified.txt
+    val rgx =
+      """(\d+).simplified.txt""".r
+
+    dir.listFiles().filter(_.getName.contains("simplified")).map { file =>
+      val rgx(numStr) = file.getName
+      numStr.toInt
+    }
+  }
+
+  private def getNextVariantNumber(name: String): Int = {
+    val existingVariants = getExistingVariants(name)
+    if (existingVariants.isEmpty) {
+      0
+    } else {
+      existingVariants.max + 1
+    }
+  }
+
+  private def getDirForTest(name: String): File = {
+    new File(goldenFilePath, s"$name")
+  }
+
+  private def isApproved(name: String, plan: SparkPlan): Boolean = {
```

Comment:
This method has 2 callers, and both of them need to get the simplified plan of `plan` again. To avoid repeated work, I think it's better to let this method takes the simplified plan as input.